### PR TITLE
Websocket auth

### PR DESCRIPTION
**Pull request by chenilim**
Added an auth op to websocket. Main structural changes:
* Moved GetSession from App into a new Auth class
* Pass Auth class to webSocket and App
* OctoListener passes session token on open

This probably needs refactoring in the future, but I can see more code going into an auth layer above app later on.

### CODE REVIEW

File: server/ws/websockets.go [hunk 1]
```diff
@@ -154,20 +109,123 @@ func (ws *Server) handleWebSocketOnChange(w http.ResponseWriter, r *http.Request
...
+
+	// Authenticated
+	wsSession.isAuthenticated = true
+	log.Printf("authenticateListener: Authenticated")
+}
+
+// AddListener adds a listener for a block's change.
+func (ws *Server) addListener(wsSession *websocketSession, blockIDs []string) {
+	if !wsSession.isAuthenticated {
+		log.Printf("addListener: NOT AUTHENTICATED")
```
Comment (jespino) [inline server/ws/websockets.go:158]: Maybe sending a response like `{"error": "NOT AUTHENTICATED"}`?

File: server/ws/websockets.go [hunk 2]
```diff
@@ -154,20 +109,123 @@ func (ws *Server) handleWebSocketOnChange(w http.ResponseWriter, r *http.Request
...
+
+		ws.listeners[key] = listeners
+	}
+	ws.mu.Unlock()
+}
+
+// removeListenerFromBlocks removes a webSocket listener from a set of block.
+func (ws *Server) removeListenerFromBlocks(wsSession *websocketSession, blockIDs []string) {
+	if !wsSession.isAuthenticated {
+		log.Printf("removeListenerFromBlocks: NOT AUTHENTICATED")
```
Comment (jespino) [inline server/ws/websockets.go:193]: Same here

File: server/ws/websockets.go [hunk 3]
```diff
@@ -154,20 +109,123 @@ func (ws *Server) handleWebSocketOnChange(w http.ResponseWriter, r *http.Request
...
+				break
+			}
+		}
+	}
+
+	ws.mu.Unlock()
+}
+
+// GetListeners returns the listeners to a blockID's changes.
+func (ws *Server) GetListeners(blockID string) []*websocket.Conn {
```
Comment (jespino) [inline server/ws/websockets.go:221]:
```suggestion
// getListeners returns the listeners to a blockID's changes.
func (ws *Server) getListeners(blockID string) []*websocket.Conn {
```
and of course is needed to change the call to `GetListeners` called later.

File: server/app/blocks_test.go [hunk 4]
```diff
@@ -13,14 +14,19 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+func isValidSessionToken(token string) bool {
```
Comment (jespino) [inline server/app/blocks_test.go:17]: Is this use anywhere?

**Review by jespino — APPROVED** [review]
a couple of tiny comments, but otherwise looks good to me.

File: server/app/blocks_test.go (hunk 4 above)
Comment (chenilim) [reply server/app/blocks_test.go]: Ah, no. Was trying a different approach. Will clean this up.

File: server/ws/websockets.go (hunk 1 above)
Comment (chenilim) [reply server/ws/websockets.go:163]: Good call. Added an error message to the client.

File: server/ws/websockets.go (hunk 3 above)
Comment (chenilim) [reply server/ws/websockets.go]: Thanks. Cleaned up the function visibility.

**Pull request merged**